src/http/request.cr

```diff
@@ -22,8 +22,45 @@ class HTTP::Request
     @cookies ||= Cookies.from_headers(headers)
   end
 
+  # Lazily parses and returns URI struct from request.
   def uri
-    URI.parse(@path)
+    (@uri ||= URI.parse(@resource)).not_nil!
+  end
```

Comment:
I wonder if this should take `@headers["Host"]` into account if present. It definitely should make sure the scheme is `"http"` I think, even though that may be misleading towards HTTPS requests.